src/libopensc/card-esteid2018.c

```diff
@@ -147,16 +143,20 @@ static int esteid_set_security_env(sc_card_t *card, const sc_security_env_t *env
 
 	sc_log(card->ctx, "algo: %lu operation: %d keyref: %d", env->algorithm, env->operation, env->key_ref[0]);
 
-	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN && env->key_ref[0] == AUTH_REF) {
-		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xA4, cse_crt_aut, sizeof(cse_crt_aut), NULL, 0);
-	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN && env->key_ref[0] == SIGN_REF) {
-		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xB6, cse_crt_sig, sizeof(cse_crt_sig), NULL, 0);
-	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_DERIVE && env->key_ref[0] == AUTH_REF) {
-		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xB8, cse_crt_dec, sizeof(cse_crt_dec), NULL, 0);
+	new_env = *env;
+	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN && (env->key_ref[0] & SIGN_REF) > 0) {
+		new_env.algorithm_ref = 0x54;
+	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN) {
+		new_env.operation = SC_SEC_OPERATION_AUTHENTICATE;
+		new_env.algorithm_ref = 0x04;
+	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_DERIVE) {
+		new_env.algorithm_ref = 0x0B;
```

Comment:
I find the long lines of repeated conditionals hard to follow. Would something like this work? Or even breaking it to some switch/case construct:
```suggestion
	if (env->algorithm == SC_ALGORITHM_EC) {
		if (env->operation == SC_SEC_OPERATION_SIGN) {
			if (env->key_ref[0] & SIGN_REF) > 0) {
				new_env.algorithm_ref = 0x54;
			} else {
				new_env.operation = SC_SEC_OPERATION_AUTHENTICATE;
				new_env.algorithm_ref = 0x04;
			}
		} else if (env->operation == SC_SEC_OPERATION_DERIVE) {
			new_env.algorithm_ref = 0x0B;
		} else {
			// error
		}
```